UPGRADE-2.0.md

```diff
@@ -0,0 +1,164 @@
+# Migration guide from Foundry 1.x to 2.0
+
+Foundry 2 has changed some of its API.
+The global philosophy is still the same.
+The main change is that we've introduced a separation between "object" factories,
+"persistence" factories and "persistence with proxy" factories.
+
+When Foundry 1.x was "persistence first", Foundry 2 is "object first".
+This would allow more decoupling from the persistence layer.
+
+## How to
+
+Every modification needed for a 1.x to 2.0 migration is covered by a deprecation.
+All you have to do is to upgrade to the last 1.x version, activate the deprecation helper, make the tests run, and
+fix all the deprecations reported.
+
+Here is an example of how the deprecation helper can be activated.
+You should set the `SYMFONY_DEPRECATIONS_HELPER` variable in `phpunit.xml` or `.env.local` file:
+```shell
+SYMFONY_DEPRECATIONS_HELPER="max[self]=0&amp;max[direct]=0&amp;quiet[]=indirect&amp;quiet[]=other"
+```
+
+Here is the full list of modifications needed:
+
+## Factory
+
+- `withAttributes()` and `addState()` are both deprecated in favor of `with()`
+- `sequence()` and `createSequence()` do not accept `callable` as a parameter anymore (how do we leverage `__index` instead??)
+
+### Change factories' parent class
+
+`Zenstruck\Foundry\ModelFactory` is now deprecated.
+You should choose between `\Zenstruck\Foundry\Persistence\PersistentProxyObjectFactory` and `\Zenstruck\Foundry\Persistence\PersistentObjectFactory`.
+Both act the same way, but the latter does not return a `Proxy` object.
+
+> [!NOTE]
+> You will have to change some methods prototypes in your classes:
+
+```php
+// before
+protected function getDefaults(): array
+{
+    // ...
+}
+
+// after
+protected function defaults(): array|callable
+{
+    // ...
+}
+```
+
+```php
+// before
+protected static function getClass(): string
+{
+    // ...
+}
+
+// after
+public static function class(): string
+{
+    // ...
+}
+```
+
+As a rule of thumb to help you to choose between these two new factory parent classes:
+- only entities (ORM) or documents (ODM) should use `PersistentProxyObjectFactory`
+- you should only use `PersistentProxyObjectFactory` if you want to leverage "auto refresh" behavior
+- all "non-persistable" objects could use `PersistentObjectFactory`, since `Proxy` is only useful for entities/documents
```

Comment:
This is a little weird - in 2.0, we'll have `ObjectFactory` for this and since this is an 1.x-2.0 upgrade guide it should show the new base class.

Couple ideas:
1. add `ObjectFactory` to 1.x (I am kind of thinking we should do this as in 2.0, you won't be able to use non-persistent objects with `PersistentObjectFactory` so would be a hard-BC break, right?)
2. rename this document `UPGRADE-1.?.md`, then add an `UPGRADE-2.0.md` in 2.x

Reply:
as I told you on slack, I'm indeed in favor to introduce `ObjectFactory` in 1.x :smile:

Reply:
https://github.com/zenstruck/foundry/pull/543

Reply:
I've re-worded this section